src/sampling.jl

```diff
@@ -0,0 +1,253 @@
+export grid_sample, ∇grid_sample
+
+@inline in_bounds(h, w, H, W) = 1 ≤ h ≤ H && 1 ≤ w ≤ W
+# Borders are considered out-of-bounds for gradient.
+@inline clip_coordinate(coordinate, dim_size) = min(dim_size, max(1, coordinate))
+@inline function ∇clip_coordinate(coordinate::C, dim_size) where C
+    if coordinate ≤ 1
+        return C(1), C(0)
+    elseif coordinate ≥ dim_size
+        return C(dim_size), C(0)
+    end
+    coordinate, C(1)
+end
+
+@inline unnormalize(coordinate, dim_size) = ((coordinate + 1.0) * 0.5) * (dim_size - 1.0) + 1.0
+@inline ∇unnormalize(coordinate, dim_size) = unnormalize(coordinate, dim_size), (dim_size - 1.0) * 0.5
+
+@inline compute_source_index(coordinate, dim_size, ::Val{:zeros}) = unnormalize(coordinate, dim_size)
+@inline compute_source_index(coordinate, dim_size, ::Val{:border}) = clip_coordinate(unnormalize(coordinate, dim_size), dim_size)
+
+@inline ∇compute_source_index(coordinate, dim_size, ::Val{:zeros}) = ∇unnormalize(coordinate, dim_size)
+@inline function ∇compute_source_index(coordinate, dim_size, ::Val{:border})
+    source_coordinate, grad_in = ∇unnormalize(coordinate, dim_size)
+    source_coordinate, grad_clip = ∇clip_coordinate(source_coordinate, dim_size)
+    source_coordinate, grad_in * grad_clip
+end
+
+"""
+    grid_sample(input::AbstractArray{T, 4}, grid::AbstractArray{T, 4}; padding_mode = Val(:zeros))
```

Comment:
```suggestion
    grid_sample(input::AbstractArray{T, 4}, grid::AbstractArray{T, 4}; padding_mode = :zeros)
```
Our API should not require users to wrap symbols in a `Val`